Add support for external infra clusters in KubeVirt platform

orenc1 opened the pull request:
This PR aims to provide an option to create guest cluster infra workloads (VMIs) on another cluster than the management cluster in which hypershift operator is running. It is done by passing a kubeconfig file path as an argument (`--external-cluster-kubeconfig-file`) to the `hypershift create cluster kubevirt`CLI command, which in turn will:
1. create a secret containing the kubeconfig
2. link it as a secret reference in the KubeVirt stanza in HostedCluster CR
3. add it to the KubeVirtCluster CR for consumption via CAPK

Signed-off-by: Oren Cohen <ocohen@redhat.com>

<!--
- Please ensure code changes are split into a series of logically independent commits.
- Every commit should have a subject/title (What) and a description/body (Why).
- Every PR must have a description.
- As an example you can use git commit -m"What" -m"Why" to achieve the requirements above. GitHub automatically recognises the commit description (-m"Why") in single commit PRs and adds it as the PR description.
- Use the [imperative mood](https://en.wikipedia.org/wiki/Imperative_mood) in the subject line for every commit. E.g `Mark infraID as required` instead of `This patch marks infraID as required` (This follows Git’s own built-in conventions). See https://github.com/openshift/hypershift/pull/485 as an example.
- See https://hypershift-docs.netlify.app/contribute for more details.

Delete this text before submitting the PR.
-->

**What this PR does / why we need it**:
Provide option to deploy kubevirt-based guest clusters on external infra clusters than the management cluster.
**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #

**Checklist**
- [ ] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

davidvossel commented on api/v1beta1/hostedcluster_types.go:
```diff
@@ -86,7 +86,9 @@ const (
 	// the customer AWS credentials in the unmanaged authentication strategy for AWS KMS secret encryption
 	AWSCredentialsFileSecretKey = "credentials"
-
+	// KubeVirtExternalInfraCredentialsSecretKey defines the kubeconfig needed for the hosted control plane operator
+	// to access and manage kubevirt workloads on an external infra cluster.
+	KubeVirtExternalInfraCredentialsSecretKey = "kubevirt-external-credentials"
```
is this used yet?

orenc1 replied:
eventually I didn't use that const. removed.

sjenning commented:
/hold

route53 rate limiting is jamming everything up and need to get https://github.com/openshift/hypershift/pull/2018 to fix it

Can remove hold after it merges.

davidvossel commented on api/v1beta1/hostedcluster_types.go:
```diff
@@ -733,6 +741,17 @@ type PowerVSPlatformSpec struct {
 }
 
+type KubeVirtPlatformSpec struct {
+	// ExtClusterName is the name of the external infra cluster used to host the guest cluster nodes
+	//
+	// +immutable
+	ExtClusterName string `json:"externalClusterName,omitempty"`
+	// KubeConfigSecret is the secret name that contains the kubeconfig for the external infra cluster
+	//
+	// +immutable
+	KubeConfigSecret *KubeconfigSecretRef `json:"kubeConfigSecret,omitempty"`
```
I'm unsure if we need `ExtClusterName`. The secret ref should have the name of the secret that holds the creds, which should be all that's necessary.

In the `example.go` file, the unique secret can be generated using the hosted cluster name and a prefix, similar to how azure generates its credential secret

So, something like this in example.go for the kubevirt secret.
```
Name:      o.Name + "-infra-credentials",
```

Then when copying that over to the hosted control plane namespace, the credential name should be in the hostedcluster.platform.kubevirt.kubeConfigSecret ref just as you have it now.

orenc1 replied:
fixed.

davidvossel commented on api/fixtures/example.go:
```diff
@@ -298,6 +300,30 @@ web_identity_token_file = /var/run/secrets/openshift/serviceaccount/token
+				ExtClusterName: o.Kubevirt.ExtClusterKubeConfig.ClusterName,
+				KubeConfigSecret: &hyperv1.KubeconfigSecretRef{
+					Name: extKubeConfigSecret.Name,
+					Key:  hostedcontrolplane.DefaultAdminKubeconfigKey,
```
`Key:  hostedcontrolplane.DefaultAdminKubeconfigKey,` - i don't understand that part.

orenc1 replied:
it is the key of the Secret object that contains the kubeconfig. the key is `kubeconfig` and it is used as a const at `hostedcontrolplane.DefaultAdminKubeconfigKey` and I reused it.
But now i removed it, for clarity.

davidvossel reviewed (commented):
Looks like a great start!

The only thing that jumped out at me a bit was the usage of the `ExtClusterName` variable. I'm not sure how relevant naming the external infra cluster is. At least it's not obvious to me why we couldn't use the HostedCluster.Name + "kubevirt-infra-cred" to create a unique per cluster secret

davidvossel commented on api/fixtures/example.go:
```diff
@@ -298,6 +300,30 @@ web_identity_token_file = /var/run/secrets/openshift/serviceaccount/token
 			panic(fmt.Sprintf("service publishing type %s is not supported", o.Kubevirt.ServicePublishingStrategy))
 		}
+		if o.Kubevirt.ExtClusterKubeConfig.KubeConfigContents != nil && o.Kubevirt.ExtClusterKubeConfig.ClusterName != "" {
+			extKubeConfigSecret := &corev1.Secret{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "Secret",
+					APIVersion: corev1.SchemeGroupVersion.String(),
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: o.Namespace,
+					Name:      o.Kubevirt.ExtClusterKubeConfig.ClusterName + "-kubeconfig",
+				},
+				Data: map[string][]byte{
+					"kubeconfig": []byte(base64.StdEncoding.EncodeToString(o.Kubevirt.ExtClusterKubeConfig.KubeConfigContents)),
+				},
+			}
```
one more comment i missed in my previous review.

The secret we copy into the hosted control plane's namespace will automatically get cleaned up when the namespace is removed.

What cleans up this secret made in the hosted cluster's namespace though?

orenc1 replied:
I'll add an implementation for that secret deletion in the kubevirt destroy cmd.

orenc1 replied:
actually, there's already a mechanism that removes the secrets from the `clusters` namespace in the destroy command, as long as they are labeled with `hypershift.openshift.io/auto-created-for-infra=InfraID`:
https://github.com/openshift/hypershift/blob/main/cmd/cluster/core/destroy.go#L149

For that purpose, I populated the InfraID with a value at:
https://github.com/openshift/hypershift/pull/2017/files#diff-a8a4a70fd459a31d15b6500359d7e0e3cd47c5f7a48f551071756ddc395eaa20R92-R96

orenc1 commented:
Hi @davidvossel ,
I addressed most of your comments, I simplified the API - removed the name field from the external kubeconfig struct and it is now just the kubeconfig contents.

However, when trying to test it e2e, CAPK was complaining it cannot unmarshal the kubeconfig.
This is the error given:
```
E0112 17:02:56.955644       1 controller.go:317] controller/kubevirtmachine "msg"="Reconciler error" "error"="failed to generate infra cluster client: failed to create REST config: couldn't get version/kind; json parse error: json: cannot unmarshal string into Go value of type struct { APIVersion string \"json:\\\"apiVersion,omitempty\\\"\"; Kind string \"json:\\\"kind,omitempty\\\"\" }" "name"="guest1-48mcs" "namespace"="clusters-guest1" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="KubevirtMachine" 
```

Looks like the error is coming from here:
https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/blob/main/pkg/infracluster/infracluster.go#L75-L78
But I was unable yet to decipher what is exactly wrong with the kubeconfig inside the secret. the contents in the secret are a valid kubeconfig file given by openshift-installer (base64-encoded).

davidvossel commented:
> However, when trying to test it e2e, CAPK was complaining it cannot unmarshal the kubeconfig.

@orenc1 It's hard for me to say why that's happening right now. What I do know is that this capk logic is being exercised by our upstream e2e test suite [here](https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/blob/884c8ca19f57c6567a8661ab244344f9b587648e/e2e/create-cluster_test.go#L901 ). Maybe that will offer some clues into why it works for the upstream kubeconfig but not the one you're testing with. 

So in that upstream test example, the kubeconfig that's being parsed is the one that's generated by kubevirt-ci. Even though we know this scenario will fail, you could take a kubeconfig generated for a kubevirtci cluster at try it with your hypershift cluster to see if the same marshaling error occurs... obviously that kubeconfig is useless, but it might give you a data point if we know it at least gets parsed correctly.

davidvossel commented on api/fixtures/example.go:
```diff
@@ -271,6 +272,30 @@ func (o ExampleOptions) Resources() *ExampleResources {
+					Namespace: o.Namespace,
+				},
+				Data: map[string][]byte{
+					"kubeconfig": []byte(base64.StdEncoding.EncodeToString(o.Kubevirt.ExtClusterKubeConfig)),
```
@orenc1 Oh, maybe this is the issue. Upstream, we didn't base64 encode the kubeconfig when this is being tested.

https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/blob/884c8ca19f57c6567a8661ab244344f9b587648e/e2e/create-cluster_test.go#L908

orenc1 replied:
yes, that was the issue. thanks.

davidvossel reviewed (commented):
One thing we probably need to start thinking about is what namespace the VMs will be running in on external infra.

Right now the VMs automatically get placed in the HCP's namespace, but on external infra that namespace won't exist. We'll either need to create a namespace, or require that a namespace is provided along with the kubeconfig to indicate where the VMs shoudl run on the external infra... We might need to consider a network policy for isolation as well for the VM pods on the external infra.

orenc1 commented:
> One thing we probably need to start thinking about is what namespace the VMs will be running in on external infra.
> 
> Right now the VMs automatically get placed in the HCP's namespace, but on external infra that namespace won't exist. We'll either need to create a namespace, or require that a namespace is provided along with the kubeconfig to indicate where the VMs shoudl run on the external infra... We might need to consider a network policy for isolation as well for the VM pods on the external infra.

The VMs on the external cluster will run at the same namespace as it was if they were created on the management cluster. I added a namespace definition for the `Kubevirtmachine.spec.template.objectmeta.namespace` at:
https://github.com/openshift/hypershift/pull/2017/files#diff-2d8fde4a95d3b1dc2de630b5b40ce849f937a940143e4b117ab32f22cec760a9R242

If the namespace does not exist, the VMs won't be created.
They will be created once the namespace is created, because the controller is looping on that.

davidvossel commented on hypershift-operator/controllers/nodepool/kubevirt/kubevirt.go:
```diff
@@ -238,6 +239,8 @@ func MachineTemplateSpec(image string, nodePool *hyperv1.NodePool) *capikubevirt
 	vmTemplate.Spec.Template.ObjectMeta.Labels[nodePoolNameLabelKey] = nodePool.Name
 	vmTemplate.ObjectMeta.Labels[nodePoolNameLabelKey] = nodePool.Name
 
+	vmTemplate.ObjectMeta.Namespace = manifests.HostedControlPlaneNamespace(nodePool.Namespace, nodePool.Name).Name
```
fyi, the `nodepool.Name` doesn't necessarily correlate to the HostedCluster's name. that could result in some unexpected results here with how the namespace is generated.

orenc1 replied:
yes, i've noticed that.
i'm now passing the hcluster object to `MachineTemplateSpec()` for its name & namespace.

orenc1 commented:
/test kubevirt-e2e-kubevirt-aws-ovn

davidvossel commented on api/v1beta1/hostedcluster_types.go:
```diff
@@ -676,6 +675,18 @@ type KubevirtPlatformSpec struct {
 	// +optional
 	// +immutable
 	BaseDomainPassthrough *bool `json:"baseDomainPassthrough,omitempty"`
+	// ExtClusterKubeConfigSecret is the secret name that holds the kubeconfig for the external infra cluster
+	//
+	// +optional
+	// +immutable
+	ExtClusterKubeConfigSecret *KubeconfigSecretRef `json:"kubeConfigSecret,omitempty"`
```
the json portion of this probably needs a more descriptive name than `kubeConfigSecret`.

orenc1 replied:
changed to `extClusterKubeConfigSecret`

davidvossel commented on cmd/cluster/kubevirt/create.go:
```diff
@@ -30,15 +32,18 @@ func NewCreateCommand(opts *core.CreateOptions) *cobra.Command {
 	cmd.Flags().Uint32Var(&opts.KubevirtPlatform.RootVolumeSize, "root-volume-size", opts.KubevirtPlatform.RootVolumeSize, "The size of the root volume for machines in the NodePool in Gi")
 	cmd.Flags().StringVar(&opts.KubevirtPlatform.ContainerDiskImage, "containerdisk", opts.KubevirtPlatform.ContainerDiskImage, "A reference to docker image with the embedded disk to be used to create the machines")
 	cmd.Flags().StringVar(&opts.KubevirtPlatform.ServicePublishingStrategy, "service-publishing-strategy", opts.KubevirtPlatform.ServicePublishingStrategy, fmt.Sprintf("Define how to expose the cluster services. Supported options: %s (Use LoadBalancer and Route to expose services), %s (Select a random node to expose service access through)", IngressServicePublishingStrategy, NodePortServicePublishin ...
+	cmd.Flags().StringVar(&opts.KubevirtPlatform.ExtClusterKubeConfigFile, "external-cluster-kubeconfig-file", opts.KubevirtPlatform.ExtClusterKubeConfigFile, "Path to a kubeconfig file of an external cluster to be used to create the guest clusters nodes onto")
```
We'll probably need to discuss these terms with the hypershift team a bit. it's possible that `external-cluster-kubeconfig-file` will need to turn into something like `infra-cluster-kubeconfig-file`. Somehow we need to converge our terminology here to match both the user facing API and our internal representation of this kubeconfig.

orenc1 replied:
updated as suggested.

davidvossel commented on control-plane-operator/controllers/hostedcontrolplane/cloud/kubevirt/providerconfig.go:
```diff
@@ -48,6 +48,7 @@ func cloudConfig(namespace string) CloudConfig {
 		},
-		Namespace: namespace,
+		Namespace:  namespace,
+		Kubeconfig: kubeconfig,
```
it's possible we shouldn't even set this path if the kubeconfig path is empty. I'm not sure how cloud provider kubevirt interprets this if the key exists without a value.

orenc1 replied:
in case that the infra cluster is local (i.e. infra cluster = mgmt cluster), the `kubeconfig` key in the `kubevirt-cloud-config` config map is an empty string, and in that case, cloud provider kubevirt uses the local cluster:
https://github.com/kubevirt/cloud-provider-kubevirt/blob/3ee828fd7530b441ca726f72bf34c3e2e88d3a9a/pkg/provider/cloud.go#L105

davidvossel replied:
perfect, thanks for looking that up

davidvossel commented on control-plane-operator/controllers/hostedcontrolplane/cloud/kubevirt/providerconfig.go:
```diff
@@ -39,7 +39,7 @@ func (c *CloudConfig) serialize() (string, error) {
 }
 
-func cloudConfig(namespace string) CloudConfig {
+func cloudConfig(namespace string, kubeconfig string) CloudConfig {
```
maybe call `kubeconfig` something like `kubeconfigPath` so it's obvious we're dealing with a path here.

orenc1 commented on control-plane-operator/controllers/hostedcontrolplane/cloud/kubevirt/reconcile.go:
```diff
@@ -127,13 +129,33 @@ func addVolumes(deployment *appsv1.Deployment) {
+			deployment.Spec.Template.Spec.Volumes,
+			emptyVolume,
+		)
+	}
```
Yes, I've updated the `podVolumeMounts()` function to add the infra kubeconfig volume mount only if external infra feature is being used.

orenc1 commented on control-plane-operator/controllers/hostedcontrolplane/kas/config.go:
```diff
@@ -90,7 +90,7 @@ func generateConfig(p KubeAPIServerConfigParams, version semver.Version) *kcpv1.
 								},
 								Defaults: podsecurityadmissionv1beta1.PodSecurityDefaults{
-									Enforce:        "privileged",
+									Enforce:        "restricted",
```
nope.
probably slipped in a rebase or something.

davidvossel reviewed (commented):
I left some comments in line.

My biggest suggestion here is that we find a way to align the terminology for this features across both the API and CLI as well as our internal implementation. I see various usages of infra, externalInfra, extInfra, infraCluster, externalCluster, etc... to all kind of describe the same thing. It would be great if we could focus on bringing consistency there. 

This consistency will help us as we pitch what we're doing to the larger hypershift team and as we document this in the future.

orenc1 commented on control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:
```diff
@@ -1584,6 +1601,27 @@ func (r *reconciler) ensureIngressControllersRemoved(ctx context.Context) (bool,
 		return false, fmt.Errorf("failed to force delete pods under openshift-ingress namespace: %w", errors.NewAggregate(errs))
 	}
 
+	// Remove ingress service and route that were created by HCCO
+	if !r.externalInfra {
+		// External infra is not being used, the ingress route and service will be deleted as a result
+		// of removal of the hosted control plane namespace
+		return false, nil
+	}
```
there will still be a branching logic in this case, because the client and the namespace are different in case of local infra and external infra.
the reasoning here was to avoid executing code that won't change the outcome.

orenc1 replied:
done.

orenc1 commented on api/v1beta1/hostedcluster_types.go:
```diff
@@ -676,6 +675,18 @@ type KubevirtPlatformSpec struct {
+	//
+	// +optional
+	// +immutable
+	ExtClusterKubeConfigSecret *KubeconfigSecretRef `json:"extClusterKubeConfigSecret,omitempty"`
+	// ExtClusterNamespace defines the namespace in the external infra cluster the virtual machines will be created at.
+	// If not specified, the namespace will be derived the same as if the hosted cluster is created at the management cluster, i.e.
+	// <hostedcluster_namespace>-<hostedcluster_name>
+	//
+	// +optional
+	// +immutable
+	ExtClusterNamespace string `json:"extClusterNamespace,omitempty"`
```
Refactored the terminology throughout the code to use exclusively `InfraNamespace` and `InfraKubeconfig` and its variants (`InfraKubeConfigFile`, `InfraKubeConfigSecret`, etc.)

davidvossel replied:
nice, that definitely helps.  For the places in the code where we use the infra client, do you think we should explicility call it the `kubevirtInfraClient`.  We don't need that distiction for the API since we're already within the kubevirt platform struct.

orenc1 commented on cmd/cluster/kubevirt/create.go:
```diff
@@ -30,16 +32,19 @@ func NewCreateCommand(opts *core.CreateOptions) *cobra.Command {
 	cmd.Flags().StringVar(&opts.KubevirtPlatform.RootVolumeAccessModes, "root-volume-access-modes", opts.KubevirtPlatform.RootVolumeAccessModes, "The access modes of the root volume to use for machines in the NodePool (comma-delimited list)")
 	cmd.Flags().StringVar(&opts.KubevirtPlatform.ContainerDiskImage, "containerdisk", opts.KubevirtPlatform.ContainerDiskImage, "A reference to docker image with the embedded disk to be used to create the machines")
 	cmd.Flags().StringVar(&opts.KubevirtPlatform.ServicePublishingStrategy, "service-publishing-strategy", opts.KubevirtPlatform.ServicePublishingStrategy, fmt.Sprintf("Define how to expose the cluster services. Supported options: %s (Use LoadBalancer and Route to expose services), %s (Select a random node to expose service access through)", IngressServicePublishingStrategy, NodePortServicePublishin ...
+	cmd.Flags().StringVar(&opts.KubevirtPlatform.ExtClusterKubeConfigFile, "infra-cluster-kubeconfig-file", opts.KubevirtPlatform.ExtClusterKubeConfigFile, "Path to a kubeconfig file of an external infra cluster to be used to create the guest clusters nodes onto")
+	cmd.Flags().StringVar(&opts.KubevirtPlatform.ExtClusterNamespace, "external-cluster-namespace", opts.KubevirtPlatform.ExtClusterKubeConfigFile, "Optional. The namespace in the external infra cluster the worker nodes will be created at. Note: the namespace should be manually created beforehand.")
```
Agreed. Terminology has been aligned. see https://github.com/openshift/hypershift/pull/2017#discussion_r1115401645 for details.

orenc1 commented on hypershift-operator/controllers/hostedcluster/internal/platform/kubevirt/kubevirt.go:
```diff
@@ -154,7 +165,30 @@ func (p Kubevirt) CAPIProviderDeploymentSpec(hcluster *hyperv1.HostedCluster, _
+		return fmt.Errorf("failed to get secret %s: %w", name, err)
+	}
+
+	target := &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Namespace: controlPlaneNamespace, Name: name.Name}}
+	_, err := createOrUpdate(ctx, c, target, func() error {
+		if target.Data == nil {
+			target.Data = map[string][]byte{}
+		}
+		for k, v := range source.Data {
+			target.Data[k] = v
+		}
+		return nil
+	})
+	return err
```
The second part of naming the credentials secret in the HCP namespace `kubevirt-infra-credentials` is reasonable, since there is only one secret of that kind in an HCP namespace.
But the credentials secret in the hosted cluster namespace (`clusters`) should be unique and contain the hcp name in its name.

I'm not quite sure how manually creating a credentials secret by the user at the `clusters` namespace with an arbitrary name (e.g. `my-super-secret`) is going to be caught and copied to the relevant HCP namespace by hypershift-operator.

davidvossel replied:
> ok, now i understood you meant the users will interact directly with the HostedCluster API and not using the hypershift CLI.
> IMHO, it makes more sense that the secrets names are aligned between clusters ns and the HCP (clusters-<guest_cluster_name>) namespace, that there will be a 1:1 relation between them.

I'm slightly worried about naming collisions in the HCP namespace. For example, say someone named their kubevirt infra secret `admin-kubeconfig` in the HC namespace. If we copied `admin-kubeconfig` directly into the HCP, then that would collide with a secret we already have in the HCP namespace that's related to the guest cluster.

orenc1 replied:
done. hard-coded the credentials secret name at the HCP namespace to `kubevirt-infra-credentials`, regardless of the original name at the hostedcluster ("`clusters`") namespace.
it is defined at:
https://github.com/openshift/hypershift/blob/d485b010e270561dda22a5a7f21645a319060dfa/api/v1beta1/hostedcluster_types.go#L138

orenc1 commented on hypershift-operator/controllers/hostedcluster/internal/platform/kubevirt/kubevirt.go:
```diff
@@ -154,7 +165,30 @@ func (p Kubevirt) CAPIProviderDeploymentSpec(hcluster *hyperv1.HostedCluster, _
+	}
+
+	var source corev1.Secret
+	name := client.ObjectKey{Namespace: hcluster.Namespace, Name: hcluster.Name + extClusterCredentialsSuffix}
```
thanks, fixed.

orenc1 commented on control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:
```diff
@@ -770,23 +774,36 @@ func (r *reconciler) reconcileIngressController(ctx context.Context, hcp *hyperv
+		if _, err := r.CreateOrUpdate(ctx, targetClient, cpService, func() error {
 			return ingress.ReconcileDefaultIngressPassthroughService(cpService, defaultIngressNodePortService, hcp)
-		}); err != nil {
+		}); err != nil && !apierrors.IsAlreadyExists(err) {
```
without ignoring the `IsAlreadyExists` error, i'm seeing a reconciliation error in the HCCO logs:
```
{"level":"error","ts":"2023-02-23T14:24:45Z","msg":"Reconciler error","controller":"resources","object":{"name":""},"namespace":"","name":"","reconcileID":"a12f1f45-ad69-4766-bbb3-77a0c9826a73","error":"[failed to reconcile ingress controller: [failed to reconcile kubevirt ingress passthrough service: services \"default-ingress-passthrough-service\" already exists, failed to reconcile kubevirt ingress passthrough route: routes.route.openshift.io \"default-ingress-passthrough-route\" already exists], MutateFn cannot mutate object name and/or object namespace. key: openshift-cluster-csi-drivers/kubevirt-csi-node-cr, kind: &TypeMeta{Kind:ClusterRole,APIVersion:rbac.authorization.k8s.io/v1,}, namespace: ]","errorCauses":[{"error":"failed to reconcile ingress controller: [failed to reconcile kubevirt ingress passthrough service: services \"default-ingress-passthrough-service\" already exists, failed to reconcile kubevirt ingress passthrough route: routes.route.openshift.io \"default-ingress-passthrough-route\" ...
```
and the ingress service & route were created successfully on the infra cluster.

davidvossel replied:
This likely has something to do with the logic around caching in the CreateOrUpdate function here [1]. We're wrapping that function with our own custom logic that seems to be tied to the mgmt cluster api somehow.

My guess is that the create or update function is using the mgmt cluster's client cache instead of the kubevirt infra client cache. Maybe the `mgr.GetAPIReader()` here [2] is using the mgmt cluster client which messes up the create or update logic.

I'm not sure exactly, but my guess is that something is off here. We might have to have our own create or update function that's configured for the kubevirt infra client specifically. 

1. https://github.com/openshift/hypershift/blob/42e85b6ed107c4b99a5ef00b886e1943ea896095/control-plane-operator/hostedclusterconfigoperator/cmd.go#L223 
2. https://github.com/openshift/hypershift/blob/42e85b6ed107c4b99a5ef00b886e1943ea896095/control-plane-operator/hostedclusterconfigoperator/cmd.go#L225

orenc1 commented on control-plane-operator/hostedclusterconfigoperator/cmd.go:
```diff
@@ -210,6 +215,22 @@ func (o *HostedClusterConfigOperator) Run(ctx context.Context) error {
 	if err := mgr.Add(cpCluster); err != nil {
 		return fmt.Errorf("cannot add CPCluster to manager: %v", err)
 	}
+	var infraConfig *rest.Config
+	var infraCluster cluster.Cluster
+	if _, err := os.Stat(o.InfraKubeconfig); err == nil {
+		infraConfig = operator.CfgFromFile(o.InfraKubeconfig)
+		infraCluster, err = cluster.New(infraConfig, func(opt *cluster.Options) {
+			opt.Namespace = o.Namespace
+			opt.Scheme = api.Scheme
+		})
+		if err != nil {
+			return fmt.Errorf("cannot create infra cluster: %v", err)
+		}
+		if err := mgr.Add(infraCluster); err != nil {
+			return fmt.Errorf("cannot add InfraCluster to manager: %v", err)
+		}
+	}
+
```
i've added an else term to default to `cpConfig` and `cpCluster` in case external kubeconfig hasn't been provided.

however, i'm not quite confident what is it going to achieve.
others could call the infra client and get the cp client instead, which could lead to unexpected results.

davidvossel replied:
> however, i'm not quite confident what is it going to achieve.
> others could call the infra client and get the cp client instead, which could lead to unexpected results.

what if we added the prefix `Kubevirt` to the InfraConfig and InfraCluster, so it's obvious this only makes sense to use in relation to kubevirt?

orenc1 replied:
ok, the `Kubevirt` prefix has been added to these reconciler variables.

orenc1 commented on control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:
```diff
@@ -770,23 +774,36 @@ func (r *reconciler) reconcileIngressController(ctx context.Context, hcp *hyperv
+		// rather than mgmt client and hosted control plane namespace
+		if hcp.Spec.Platform.Kubevirt.ExtClusterKubeConfigSecret != nil &&
+			hcp.Spec.Platform.Kubevirt.ExtClusterNamespace != "" {
+			r.externalInfra = true
+			r.externalInfraNamespace = hcp.Spec.Platform.Kubevirt.ExtClusterNamespace
```
to set these properties in the `reconciler` initialization, i need access to the HCP object at the `Setup()` function:
https://github.com/openshift/hypershift/blob/a1c5b4cb4fc86dc0bd87bd283d4ece27693112fb/control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go#L111
To retrieve the HCP from the cluster i need to set up a client there and then use the hcp name and its namespace, as done at the beginning of the `Reconcile()` function:
https://github.com/openshift/hypershift/blob/main/control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go#L206

But setting something very specific at the beginning of the Reconcile loop might not be the best place to do so. WDYT?

davidvossel replied:
maybe you don't need the `r.externalInfra` and `r.externalInfraNamespace` values on the `reconcile` object.

For example, what would it hurt for the `ensureIngressControllersRemoved` function to always attempt to use the infraClient to remove the service and route? That's part of what always having the infraClient present (regardless if it's actually the cpClient) allows us to do without needing branching logic.

Also, you can pass the hcp to the `ensureIngressControllersRemoved` function in order to get values off the hcp if you need, which also make it redundant to need to cache the `r.externalInfra` and `r.externalInfranamespace` values on the reconcile object

orenc1 replied:
done.
i was able to remove the recociler variables and consolidate some logic paths.

orenc1 commented on control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:
```diff
@@ -1584,6 +1601,27 @@ func (r *reconciler) ensureIngressControllersRemoved(ctx context.Context) (bool,
+	cpService := manifests.IngressDefaultIngressPassthroughService(r.externalInfraNamespace)
+	cpPassthroughRoute := manifests.IngressDefaultIngressPassthroughRoute(r.externalInfraNamespace)
+
+	if err := r.infraClient.Delete(ctx, cpService); err != nil {
+		errs = append(errs, fmt.Errorf("failed to delete %s: %w", client.ObjectKeyFromObject(cpService).String(), err))
+	}
+
+	if err := r.infraClient.Delete(ctx, cpPassthroughRoute); err != nil {
+		errs = append(errs, fmt.Errorf("failed to delete %s: %w", client.ObjectKeyFromObject(cpPassthroughRoute).String(), err))
+	}
+
```
right.
ignoring the case the service and/or route that are due to be deleted do not exist.

orenc1 replied:
(code updated, add `&& !apierrors.IsNotFound(err)` clause for both)

orenc1 commented on control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:
```diff
@@ -770,23 +774,36 @@ func (r *reconciler) reconcileIngressController(ctx context.Context, hcp *hyperv
-		hcpNamespace := hcp.Namespace
+		var targetClient client.Client
+		var namespace string
+		// If external infra cluster is being used - use infra client and external cluster namespace
+		// rather than mgmt client and hosted control plane namespace
+		if hcp.Spec.Platform.Kubevirt.ExtClusterKubeConfigSecret != nil &&
+			hcp.Spec.Platform.Kubevirt.ExtClusterNamespace != "" {
+			r.externalInfra = true
+			r.externalInfraNamespace = hcp.Spec.Platform.Kubevirt.ExtClusterNamespace
```
the problem with that is the fact that the source of `externalInfraNamespace` is in the HCP object, and when HCCO is initialized the HCP object is not yet visible to it.
therefore i set that value when it was needed and the hcp is available to consume.

regarding always use `r.infraClient` - yes, but we will still have the branching because of the target namespace detection.

davidvossel replied:
i left a more detailed follow for this on another comment. We can probably avoid the `r.externalInfra` and `r.externalInfranamespace` vars by inspecting the hcp object during creation and deletion

orenc1 replied:
done.

orenc1 commented:
/test kubevirt-e2e-kubevirt-aws-ovn

orenc1 commented:
/test kubevirt-e2e-kubevirt-aws-ovn

orenc1 commented:
/test kubevirt-e2e-kubevirt-aws-ovn

davidvossel commented on control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:
```diff
@@ -137,6 +138,23 @@ func Setup(opts *operator.HostedClusterConfigOperatorConfig) error {
 		return fmt.Errorf("failed to create uncached client: %w", err)
 	}
 
+	var kubevirtInfraClient client.Client
+	if opts.KubevirtInfraConfig != opts.Config {
+		// if kubevirt infra config is not used, it is being set as the mgmt config
+		kubevirtInfraClient, err = client.New(opts.KubevirtInfraConfig, client.Options{
+			Scheme: opts.Manager.GetScheme(),
+			Mapper: opts.Manager.GetRESTMapper(),
+			Opts: client.WarningHandlerOptions{
+				SuppressWarnings: true,
+			},
+		})
+		if err != nil {
+			return fmt.Errorf("failed to create kubevirt infra uncached client: %w", err)
+		}
```
i'd suggest always creating the kubevirtInfraClient rather than trying to reuse a client if we detect that the ```opts.KubevirtInfraConfig``` matches another known client config.

orenc1 replied:
code updated.

orenc1 commented on control-plane-operator/hostedclusterconfigoperator/controllers/resources/manifests/ingress.go:
```diff
@@ -35,19 +35,19 @@ func IngressDefaultIngressNodePortService() *corev1.Service {
 	return &routev1.Route{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      "default-ingress-passthrough-route",
+			Name:      hcpName + "-default-ingress-passthrough-route",
```
rebased on #2092 and updated the code.

orenc1 commented on control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:
```diff
@@ -137,6 +138,23 @@ func Setup(opts *operator.HostedClusterConfigOperatorConfig) error {
+		}
+	} else {
+		kubevirtInfraClient = uncachedClient
+	}
```
code updated.

orenc1 commented:
/test kubevirt-e2e-kubevirt-aws-ovn

orenc1 commented:
/test kubevirt-e2e-kubevirt-aws-ovn

orenc1 commented:
/retest

orenc1 commented:
/retest

orenc1 commented:
/test kubevirt-e2e-kubevirt-aws-ovn

davidvossel commented on control-plane-operator/controllers/hostedcontrolplane/configoperator/reconcile.go:
```diff
@@ -229,7 +230,8 @@ var (
 )
 
-func ReconcileDeployment(deployment *appsv1.Deployment, image, hcpName, openShiftVersion, kubeVersion string, ownerRef config.OwnerRef, config *config.DeploymentConfig, availabilityProberImage string, enableCIDebugOutput bool, platformType hyperv1.PlatformType, apiInternalPort *int32, konnectivityAddress string, konnectivityPort int32, oauthAddress string, oauthPort int32, releaseImage string, ad ...
+func ReconcileDeployment(deployment *appsv1.Deployment, image, hcpName, openShiftVersion, kubeVersion string, ownerRef config.OwnerRef, config *config.DeploymentConfig, availabilityProberImage string, enableCIDebugOutput bool, platformType hyperv1.PlatformType, apiInternalPort *int32, konnectivityAddress string, konnectivityPort int32, oauthAddress string, oauthPort int32, releaseImage string, ad ...
```
I'm going to recommend a refactor here in this ReconcileDeployment function that keeps our kubevirt platform codepaths completely independent of the standard HCCO deployment.

Take a look at how the ```if additionalTrustBundle != nil``` functionality. That injects volumes and volumemounts into the deployment when that setting is present. I'd like to see something similar to that for the kubevirt platform that injects our logic into the HCCO deployment after the standard HCCO deployment is already rendered.


so, something like this, right around where the additionalTrustBundle logic is
```
	if additionalTrustBundle != nil {
		util.DeploymentAddTrustBundleVolume(additionalTrustBundle, deployment)
	}

        if isExternalInfraKv(hcp) {
                // injects the kubevirt credentials secret volume, volume mount path, and appends cli arg. 
                util.DeploymentAddKubevirtInfraCredentials(deployment)
        }

```

davidvossel commented on control-plane-operator/controllers/hostedcontrolplane/configoperator/reconcile.go:
```diff
@@ -352,6 +364,19 @@ func buildHCCVolumeKubeconfig(v *corev1.Volume) {
+		v.Secret = &corev1.SecretVolumeSource{
+			SecretName: hyperv1.KubeVirtInfraCredentialsSecretName,
+		}
+	} else {
+		v.VolumeSource = corev1.VolumeSource{
+			EmptyDir: &corev1.EmptyDirVolumeSource{},
+		}
+	}
```
I think we can remove this empty dir now if we do the refactor.

davidvossel commented on control-plane-operator/controllers/hostedcontrolplane/configoperator/reconcile.go:
```diff
@@ -219,6 +219,7 @@ var (
 	volumeMounts = util.PodVolumeMounts{
 		hccContainerMain().Name: util.ContainerVolumeMounts{
 			hccVolumeKubeconfig().Name:      "/etc/kubernetes/kubeconfig",
+			hccVolumeInfraKubeconfig().Name: "/etc/kubernetes/infra-kubeconfig",
```
I think we should add the term`kubevirt` somewhere here in the path. `/etc/kubernetes/kubevirt-infra-kubeconfig`

davidvossel commented on control-plane-operator/controllers/hostedcontrolplane/configoperator/reconcile.go:
```diff
@@ -317,6 +326,9 @@ func buildHCCContainerMain(image, hcpName, openShiftVersion, kubeVersion string,
 		if platformType == hyperv1.IBMCloudPlatform {
 			c.Command = append(c.Command, "--controllers=controller-manager-ca,resources,inplaceupgrader,drainer,hcpstatus")
 		}
+		if externalInfra {
+			c.Command = append(c.Command, fmt.Sprintf("--infra-kubeconfig=%s", path.Join(volumeMounts.Path(c.Name, hccVolumeInfraKubeconfig().Name), kas.KubeconfigKey)))
+		}
```
this append could be added to the refactor function that isolates all the kubevirt logic.

davidvossel commented on control-plane-operator/controllers/hostedcontrolplane/configoperator/reconcile.go:
```diff
@@ -283,6 +286,12 @@ func hccVolumeKubeconfig() *corev1.Volume {
 
+func hccVolumeInfraKubeconfig() *corev1.Volume {
+	return &corev1.Volume{
+		Name: "infra-kubeconfig",
```
Name: "kubevirt-infra-kubeconfig"

davidvossel commented on control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:
```diff
@@ -3338,14 +3338,24 @@ func (r *HostedControlPlaneReconciler) reconcileCloudControllerManager(ctx conte
+		r.Log.Info("creating kubevirt cloud-config ConfigMap")
 		if _, err := createOrUpdate(ctx, r, ccmConfig, func() error {
-			return kubevirt.ReconcileCloudConfig(ccmConfig, hcp)
+			var infraClusterNamespace string
+			var infraKubeconfigPath string
+			if hcp.Spec.Platform.Kubevirt.InfraNamespace != "" {
+				infraClusterNamespace = hcp.Spec.Platform.Kubevirt.InfraNamespace
+				infraKubeconfigPath = "/etc/kubernetes/infra-kubeconfig/kubeconfig"
+			}
```
all this can be moved into the `kubevirt.ReconcileCloudConfig()` function i believe.

davidvossel commented on control-plane-operator/hostedclusterconfigoperator/cmd.go:
```diff
@@ -210,6 +215,14 @@ func (o *HostedClusterConfigOperator) Run(ctx context.Context) error {
 	if err := mgr.Add(cpCluster); err != nil {
 		return fmt.Errorf("cannot add CPCluster to manager: %v", err)
 	}
+	var kubevirtInfraConfig *rest.Config
+	if _, err := os.Stat(o.InfraKubeconfig); err == nil {
+		kubevirtInfraConfig = operator.CfgFromFile(o.InfraKubeconfig)
+	} else {
+		// in case infra kubeconfig hasn't been provided, default the kubevirtInfraCluster to cpConfig
+		kubevirtInfraConfig = cpConfig
+	}
+
```
this should probably read slightly differently.

as it's written right now, by not inspecting if InfraKubeconfig is set when a stat error potentially occurs, we might default to the cpConfig even when o.InfraKubeconfig is set.

I think something like this resolves the potential issue. 
```
if o.InfraKubeconfig != "" {
                kubevirtInfraConfig = operator.CfgFromFile(o.InfraKubeconfig)
} else {
		// in case infra kubeconfig hasn't been provided, default the kubevirtInfraCluster to cpConfig
		kubevirtInfraConfig = cpConfig
}

```

orenc1 replied:
the idea was to verify the kubeconfig file actually exists in the file system before attempting to construct a client from it.
anyway, i'm changing as per your suggestion.

davidvossel commented on control-plane-operator/hostedclusterconfigoperator/cmd.go:
```diff
@@ -123,6 +127,7 @@ func newHostedClusterConfigOperatorCommand() *cobra.Command {
 	flags.AddGoFlagSet(flag.CommandLine)
 	flags.StringVar(&cpo.Namespace, "namespace", cpo.Namespace, "Namespace for control plane components on management cluster")
 	flags.StringVar(&cpo.TargetKubeconfig, "target-kubeconfig", cpo.TargetKubeconfig, "Kubeconfig for target cluster")
+	flags.StringVar(&cpo.InfraKubeconfig, "infra-kubeconfig", cpo.InfraKubeconfig, "Kubeconfig for infra cluster")
```
this needs to be kubevirt specific. `kubevirt-infra-kubeconfig`

davidvossel commented on control-plane-operator/hostedclusterconfigoperator/cmd.go:
```diff
@@ -68,6 +69,9 @@ type HostedClusterConfigOperator struct {
 	TargetKubeconfig string
 
+	// InfraKubeconfig is a kubeconfig to access the infra cluster.
+	InfraKubeconfig string
```
KubevirtInfraKubeconfig

davidvossel commented on control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:
```diff
@@ -1622,6 +1641,37 @@ func (r *reconciler) ensureIngressControllersRemoved(ctx context.Context) (bool,
+			} else {
+				namespace = hcp.Namespace
+			}
+			cpService := manifests.IngressDefaultIngressPassthroughService(namespace)
+			cpPassthroughRoute := manifests.IngressDefaultIngressPassthroughRoute(namespace)
+
```
note, that with my pr, you'll need to add the GenerateID to the names here in order to locate the service and route. Take a look at my creation logic.

davidvossel commented on hypershift-operator/controllers/manifests/manifests.go:
```diff
@@ -62,3 +64,22 @@ func IngressDefaultIngressController() *operatorv1.IngressController {
+func DummyRoute(namespace string) *routev1.Route {
+	return &routev1.Route{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "dummy",
```
can we give this a slightly more descriptive name, just in case someone encounters it.

something like `kubevirt-infra-temp-route`?

orenc1 commented:
Hi @davidvossel ,
Your comments have been addressed.

/test kubevirt-e2e-kubevirt-aws-ovn

orenc1 commented:
#2230 is required for the CI, but should be removed when using release payload outside of openshift-ci

/test kubevirt-e2e-kubevirt-aws-ovn

davidvossel commented on hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:
```diff
@@ -4326,33 +4328,59 @@ func (r *HostedClusterReconciler) reconcileKubevirtPlatformDefaultSettings(ctx c
+			// kubevirtInfraTempRoute is used to resolve the base domain of the infra cluster without accessing IngressController
+			kubevirtInfraTempRoute := manifests.KubevirtInfraTempRoute(namespace)
+
+			if err := targetClient.Create(ctx, kubevirtInfraTempRoute); err != nil {
+				return err
+			}
+			host := kubevirtInfraTempRoute.Spec.Host
+			if host != "" {
+				hostParts := strings.Split(host, ".")
+				baseDomain := strings.Join(hostParts[1:], ".")
+
+				// For the KubeVirt platform, the basedomain can be autogenerated using
+				// the *.apps domain of the management/infra hosting cluster
+				// This makes the guest cluster's base domain a subdomain of the
+				// hypershift infra/mgmt cluster's base domain.
+				//
+				// Example:
+				//   Infra/Mgmt cluster's DNS
+				//     Base: example.com
+				//     Cluster: mgmt-cluster.example.com
+				//     Apps:    *.apps.mgmt-cluster.example.com
+				//   KubeVirt Guest cluster's DNS
+				//     Base: apps.mgmt-cluster.example.com
+				//     Cluster: guest.apps.mgmt-cluster.example.com
+				//     Apps: *.apps.guest.apps.mgmt-cluster.example.com
+				//
+				// This is possible using OCP wildcard routes
+				hc.Spec.DNS.BaseDomain = baseDomain
+
+				if err := targetClient.Delete(ctx, kubevirtInfraTempRoute); err != nil {
+					return err
+				}
+			}
```
The conditional logic `if host != ""` makes this code look a little problematic. I know the logic works at the moment, but the way it's structured put's us slightly at risk. Here's why

If somehow the route's `host == ""`, then the temp route will not be deleted, the BaseDomain will remain unset, this function will silently move on. However, on the next reconcile, we'd wind up right back here again because BaseDomain is still empty. Except this time the `Create` will fail because the route already exists, and we'll continue to permanently fail the reconcile from that point because the logic that cleans up the temp route will never execute. The error will make it look like we're failing to "create" something, when in reality the error should have been that we couldn't detect a host value.

I think two simple changes would help here.

1. change the Create() logic for the temp route to use the CreateOrUpdate() logic instead. I know this involves passing in that createOrUpdate function pointer through a few layers of functions, but it shouldn't be too big of a deal.
2. return an error if `host == ""`something like ... "unable to autodetect kubevirt platform BaseDomain, temporary route is not giving a host value"

With those two changes, we'll at least be bubbling up the right error if this (route with empty host) does somehow occur in the future. And it also gives the error a chance to resolve if the route get's the `host` value lazily applied later on for some reason.

orenc1 replied:
done, now using the `createOrUpdate` function for the temp route.

orenc1 replied:
done, updated the function to use `CreateOrUpdate` from `upsert` instead of the regular controller runtime `Create`

davidvossel commented on hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:
```diff
@@ -4326,33 +4328,59 @@ func (r *HostedClusterReconciler) reconcileKubevirtPlatformDefaultSettings(ctx c
+		var namespace string
+		if hc.Spec.Platform.Kubevirt.Credentials != nil {
+			var err error
+			targetClient, err = generateInfraClusterClient(ctx, r.Client, *hc)
```
maybe for clarity, this function should be renamed from generateInfraClusterClient to generateKubevirtInfraClusterClient()

davidvossel commented on hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:
```diff
@@ -4414,3 +4442,41 @@ func (r *HostedClusterReconciler) dereferenceAWSRoles(ctx context.Context, roles
 
 	return nil
 }
+
+// ClientFactoryFunc defines the function to create a new client
+type ClientFactoryFunc func(config *rest.Config, options client.Options) (client.Client, error)
```
I think this function type might not be used anymore.

davidvossel reviewed (commented):
looks great. I caught some really minor things. I think we're good to go once these couple of items are resolved.

orenc1 commented:
/test kubevirt-e2e-kubevirt-aws-ovn

davidvossel commented:
/test kubevirt-e2e-kubevirt-aws-ovn

nodes on the latest kubevirt test run never came online. It looks like the network never became available for some reason.

```
    message: 'container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady
      message:Network plugin returns error: No CNI configuration file in /etc/kubernetes/cni/net.d/.
      Has your network provider started?'
```

davidvossel reviewed (commented):
/lgtm
/approve
/hold

remove the hold once our presubmit passes

orenc1 commented:
/test kubevirt-e2e-kubevirt-aws-ovn

orenc1 commented:
cluster bootstrapping issues
/test kubevirt-e2e-kubevirt-aws-ovn

orenc1 commented:
/test kubevirt-e2e-kubevirt-aws-ovn

orenc1 commented:
/test kubevirt-e2e-kubevirt-aws-ovn

orenc1 commented:
/test kubevirt-e2e-kubevirt-aws-ovn

orenc1 commented:
/test kubevirt-e2e-kubevirt-aws-ovn

orenc1 commented:
the 2nd guest node didn't came ready due to an error in `ovnkube-node`:
```
failed to start default network controller: error looking up gw interface: "br-ex", error: Link not found
```
will try to get journal logs next run
/test kubevirt-e2e-kubevirt-aws-ovn

orenc1 commented:
ci/prow/kubevirt-e2e-kubevirt-aws-ovn passed.
/hold cancel
